src/app/queries/plans.ts

```diff
@@ -136,35 +140,111 @@ export const usePatchPlanMutation = (
+            hook.cr
+          );
+          return { ref: nameAndNamespace(response.data.metadata), instance: hook.instance };
+        }
+        const response = await client.create<IHook>(hookResource, hook.cr);
+        return { ref: nameAndNamespace(response.data.metadata), instance: hook.instance };
+      });
+
+      const hooksRef = await Promise.all(updatedHooksRef);
+
+      const updatedPlan = generatePlan(forms, networkMappingRef, storageMappingRef, hooksRef);
       const [, , planResponse] = await Promise.all([
         networkMapping &&
           client.patch<Mapping>(networkMapResource, networkMappingRef.name, networkMapping),
         storageMapping &&
           client.patch<Mapping>(storageMapResource, storageMappingRef.name, storageMapping),
+
         client.patch<IPlan>(planResource, planBeingEdited.metadata.name, updatedPlan),
       ]);
 
-      // TODO handle hooks in the API
-      //  - for each instance in forms.hooks.values.instances:
-      //    - if there is a prefilledFromHook property in the instance, patch that existing hook using generateHook
-      //    - if not, create a new hook using generateHook
-      //  - for each hook reference in the plan (plan.spec.vms[].hooks[].hook) find any that don't match instances in the form data
-      //    - those were removed in the wizard, so delete the matching hook CRs
+      // Patch new hooks with ownerReferences to the edited plan
+      const hooksWithOwnerRef: IHook[] = [];
+      for (let i = 0; i < hooksRef.length; i++) {
+        if (!hooksRef[i].instance.prefilledFromHook) {
+          hooksWithOwnerRef.push(
+            generateHook(hooksRef[i].instance, hooksRef[i].ref, '', planBeingEdited)
+          );
+        }
+      }
+      const newHooksWithOwnerRef = hooksWithOwnerRef.map(async (hookWithOwnerRef) => {
+        const response = await client.patch<IHook>(
+          hookResource,
+          (hookWithOwnerRef.metadata as IMetaObjectMeta).name,
+          hookWithOwnerRef
+        );
+        return response;
+      });
+      await Promise.all(newHooksWithOwnerRef);
+
+      // Delete hooks removed from plan
+      const planHooksToDelete = planBeingEdited.spec.vms[0].hooks?.filter((vmsHook) => {
+        if (
+          forms.hooks.values.instances.length === 0 ||
+          forms.hooks.values.instances.filter((hook) => {
+            if (
+              !hook.prefilledFromHook ||
+              (hook.prefilledFromHook.metadata as IMetaObjectMeta).name === vmsHook.hook.name
+            ) {
+              return true;
+            }
+            return false;
+          }).length === 0
+        ) {
+          return true;
+        }
+        return false;
+      });
+      const hooksToDelete = hooks.data?.items.filter((hook) =>
+        planHooksToDelete?.find((vmHook) => isSameResource(vmHook.hook, hook.metadata) || null)
+      );
+      hooksToDelete?.forEach((hook) => deleteHook(hook));
```

Comment:
I see the `deleteHook` function comes from your `useDeleteHookMutation` above, but that means it is asynchronous. We're not doing anything with the status object from it, and this plan mutation will resolve successfully even if the deletion fails. We want to make this deletion synchronously as part of the plan mutation, so that if there is any error it will bubble up and be shown in the `ResolvedQuery` component just like if something failed with the plan.

I think the solution here is to entirely get rid of `useDeleteHookMutation` and instead make the `client.delete(...)` calls right here and await them in a `Promise.all`, just like you did for patching the hooks above. And since you can do this using only the name/namespace ref, you don't even need the `const hooksToDelete`, you can just map directly over the `planHooksToDelete` from above.

Reply:
Yes that makes total sense,
I think I wrongly went on `useDeleteHook` road because of the way delete happens in other parts.
That's overkill as we just need to be synchronous
So now using `client.delete()` directly.